[MIG][12.0] port pos_order_mgmt

PierrickBrun opened the pull request:
With modules split as discussed in #334

chienandalu committed:
[11.0][ADD] pos_order_mgmt: New module

legalsylvain committed:
Update ROADMAP.rst

OCA-git-bot committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update pos_order_mgmt.pot

chienandalu committed:
[FIX] pos_order_mgmt: reprint tickets with fp

When the PoS is using fiscal positions it can lead to errors when
computing the reprinted ticket value because the tax mapping function
computes them from the current order instead of from the one belonging
to the line. So we set our loaded order temporarily as the selected one
just to bring it back to the current one after it's printed.

oca-travis committed:
[UPD] Update pos_order_mgmt.pot

mvrodriguez committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (20 of 20 strings)

Translation: pos-11.0/pos-11.0-pos_order_mgmt
Translate-URL: https://translation.odoo-community.org/projects/pos-11-0/pos-11-0-pos_order_mgmt/es/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: pos-11.0/pos-11.0-pos_order_mgmt
Translate-URL: https://translation.odoo-community.org/projects/pos-11-0/pos-11-0-pos_order_mgmt/

pedrobaeza commented:
A migration script will be needed that installs the 2 extra modules.

legalsylvain commented on pos_order_mgmt/readme/USAGE.rst:
```diff
@@ -0,0 +1,19 @@
+Once the PoS is loaded, you'll find a shopping trolley icon (🛒) in the top
+bar that grants access to the order list screen.
+
+.. image:: /pos_order_mgmt/static/description/order-mgmt-icon.png
```
replace by 
``.. image:: ../static/description/order-mgmt-icon.png``
With that syntax, 

- it will work locally in github interface if you see the USAGE.rst file in your browser
- and the main readme of the module will be correctly generated (ref : https://github.com/OCA/maintainer-tools/pull/397)

Commit:
[MIG] 12.0 with integration of pos_return

Commit:
[IMP] configure reprint and return separately

PierrickBrun commented:
after trying to split it into several modules, it appeared to be more time consuming than I first thought.

I've migrated the module without splitting but I modified the pos.config to allow to choose to allow reprint or returning done orders.

I think the returning part is working well but I' don't know the 11.0 module and pos_order_return enough to be sure about it.

legalsylvain commented on pos_order_mgmt/static/src/js/widgets.js:
```diff
@@ -151,34 +151,40 @@ odoo.define('pos_order_mgmt.widgets', function (require) {
-                this.pos.reloaded_order = order;
-                this.gui.show_screen('receipt');
-                this.pos.reloaded_order = false;
+            //this is only a double check, the button is invisible
```
why adding a double check, if the button is invisible ?

legalsylvain reviewed (commented):
> after trying to split it into several modules, it appeared to be more time consuming than I first thought.

> I've migrated the module without splitting but I modified the pos.config to allow to choose to allow reprint or returning done orders.

Thanks @PierrickBrun for porting this module.

> I think the returning part is working well but I' don't know the 11.0 module and pos_order_return enough to be sure about it.

Well, ``pos_order_return`` was developed to 
- link returned order to the original order, (what your code does)
- to block the possibility  to return more quantity as the original one. This feature was developped only in backoffice part and had to be implemented in the front office.
As in new version of Odoo PoS, it is not possible to make an order in a backoffice part, we could consider ``pos_order_return`` as fully obsolete on the second point.

2. Did you checked if the module is compatible with ``pos_restaurant`` ? ref : https://github.com/OCA/pos/issues/333. CC @chienandalu, @carlosDomatix 

kind regards.

other wise, LGTM. (code review / no test).

PierrickBrun commented:
Looks like it works better with pos_restaurant now

Commit:
[IMP] Add Duplicate info on reprinted receipt

Commit:
[IMP] allow srucharge of orderlines loading

fuentes73 reviewed (approved):
Tested and working OK

fuentes73 commented on pos_order_mgmt/static/src/js/widgets.js:
```diff
@@ -0,0 +1,404 @@
+                if (line.length === 3) {
+                    line = line[2];
+                }
+                var product = self.pos.db.get_product_by_id(line.product_id);
```
Is missing at the beginning `var self = this;`
So this line throws the error: `Cannot read property 'db' of undefined`

fuentes73 commented on pos_order_mgmt/static/src/js/widgets.js:
```diff
@@ -0,0 +1,404 @@
+                // We'll refunded orders once they are synced
+                order.returned_order_id = order_data.id || order_data.name;
+                order.origin_name = order_data.pos_reference || order.returned_order_id;
+                return order;
```
Testing more in depth the module I also seen that, if you return and order and reload the page without change anything, the "return" info is lost.
![return-pos-order](https://user-images.githubusercontent.com/4355395/57447953-91c7f500-7258-11e9-9523-2c968139be4c.gif)


A possible solution might be to add `order.trigger('change');` before return the order
```suggestion
                order.trigger('change');
                return order;
```

PierrickBrun commented:
thank you for the review @fuentes010 , I've addressed the changes

Kerrim- commented on pos_order_mgmt/static/src/js/widgets.js:
```diff
@@ -0,0 +1,406 @@
+                    });
+                    correct_order_print = false;
+                }
+                if (correct_order_print && action === 'print') {
```
Hi,

I'm using this code for a project and wanted to ask if it is possible to change this part to be a little easier to extend. For example by putting it in some kind of dispatch method or like this:

```suggestion
                if(correct_order_print){
                    self["action_" + action](order);
                }
```

**Pros**: Easier to extend (You only need to register the event and write the method)
**Cons**: If you need different checks before executing the action or you want to use actions not starting with action this suggestion won't work. (But the dispatch method still would.)

Kerrim- replied:
Also if there is more work to do with this module I would be glad to help :)

PierrickBrun replied:
Hi @Kerrim- , No problem about accepting a PR from my side :+1: I only use the module for reprinting tickets so I did not dig a lot into returns

pedrobaeza commented:
Please add #354 here

ivantodorovich commented:
Functional test only.

I only encounter this issue (very small one):
On 11.0. the order list window closes automatically when clicking the refund order button.

**12.0:**
![9Lx4sxJ4Rw](https://user-images.githubusercontent.com/1914185/59455289-24d5eb00-8e14-11e9-97cf-5733f5fb7857.gif)


**11.0:**
![H4VJPthjQu](https://user-images.githubusercontent.com/1914185/59455294-27d0db80-8e14-11e9-9447-8626d6373868.gif)

ivantodorovich commented:
It's probably because of this line:
https://github.com/OCA/pos/pull/337/commits/e548fccd518bf2653723a1107d058b50a41dcdb5?file-filters%5B%5D=.js&owned-by%5B%5D=#diff-3619c2549c163cd99597081619ddb2c4R273

PierrickBrun commented:
@ivantodorovich nice catch :+1:

ivantodorovich committed:
- Make search_query a property
- Fix search clear button

- Refactor load_order and actions

This refactors load_data into two functions: load_order_data and load_order_from_data.
Also, when executing actions, instead of sending the order as an argument, only the order_data will be sent.
The action method is responsible for generating the order data (if that's what it needs)

All the refund logic has been extracted from the various methods, and it's now in action_return.

Now, returns have a specific 'Refund XXXX' name.

- IMP Allow to duplicate done orders from the POS

- IMP filter by customer name, too, when searching

PierrickBrun committed:
Merge pull request #1 from druidoo/12-order_mgmt_imp

12 order mgmt imp

chienandalu commented:
Hi, @PierrickBrun , maybe you want to take a look at this: https://github.com/OCA/pos/pull/361

legalsylvain committed:
[FIX] loaded orders date to local https://github.com/OCA/pos/pull/361 cortesy @chienandalu
[FIX] compatibility with pos_restaurant https://github.com/OCA/pos/pull/339 cortesy @carlosDomatix
[FIX] add _compute_refund_order_qty
[ADD] pos order view form. Add links to original returned order and refund orders.
[IMP] when refunding an order via the 'Refund' button in back office, set the correct returned_order_id.
[IMP] Add helper on each button
[ADD] python test
[REF] JS remove useless code
[REF] Python remove useless code
[FIX] remove bad oldname
[FIX] remove gap in bill display, when it is a refund
[FIX] when reprinting a bill, the returned order name is displayed, if any
[FIX] JS lint, with OCA eslint file
[ADD] french translation

legalsylvain commented on pos_order_mgmt/models/pos_order.py:
```diff
@@ -0,0 +1,131 @@
+        readonly=True,
+    )
+    refund_order_qty = fields.Integer(
+        compute='_compute_refund_order_qty',
```
Where is this function defined ?

legalsylvain commented on pos_order_mgmt/static/src/js/widgets.js:
```diff
@@ -0,0 +1,450 @@
+                })
+            } else {
+                _.each(this.orders, function (order_data) {
+                    if (order.uid === dataset.uid) {
```
I don't understand in which case dataset.orderId is not defined, and order is not defined anywhere.

legalsylvain reviewed (changes requested):
Hi @PierrickBrun. 

* some minor errors inline. 
* there are also a lot of lint warning on Travis. 
* Coverage is low because there is no python test
* Adding a ribbon is cool, but generate a display bug in the total place : 
![image](https://user-images.githubusercontent.com/3407482/61219457-8a1f4380-a714-11e9-801e-9360a0c392c3.png)




I propose to fix all that points this week, in a PR against your branch.

Otherwise, thanks a lot for this work ! 
kind regards.

legalsylvain commented on pos_order_mgmt/models/pos_config.py:
```diff
@@ -0,0 +1,40 @@
+        help='Allows to duplicate already done orders in the frontend',
+    )
+
+    iface_load_done_order_max_qty = fields.Integer(
```
This field is not used anywhere.

legalsylvain commented on pos_order_mgmt/models/pos_config.py:
```diff
@@ -0,0 +1,40 @@
+        string='Return Done Orders',
+        default=True,
+        help='Allows to return already done orders in the frontend',
+        oldname='iface_load_done_order',
```
there are two field with the same oldname value. 'iface_load_done_order'. why ?

PierrickBrun replied:
the field iface_load_done_order is splitted, so if you used it, you want both the functions once migrated

legalsylvain commented:
Hello, I finished the port of this module, fixing some points, and adding test. (coverage from 43% to 71%).
I did a PR against this PR : https://github.com/OCA/pos/pull/373
You can test it on the runbot of the PR.

regards.

PierrickBrun committed:
Merge pull request #3 from legalsylvain/12.0-IMP-mig-pos_order_mgmt

[12][IMP] port of pos_order_mgmt

legalsylvain commented:
@fuentes010, @ivantodorovich : various changes since your last review. do you want to update your review ? 
thanks.

rafaelbn reviewed (changes requested):
Please don't merge until @chienandalu  make last review tomorrow, thanks!

chienandalu reviewed (approved):
Tested on runbot and code review. Thanks for the good work!

legalsylvain commented:
Hi @chienandalu, thanks for your review.

/ocabot merge
